stellargraph/random.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import random
+import tensorflow as tf
+
+
+_sg_random = random.Random()
+
+
+def generate_seed(seed=None):
+    """
+    Convenience function to use the global seed by default if the provided seed is None.
+
+    Args:
+        seed (int, optional): seed value
+
+    Returns:
+        seed if not None, otherwise the global seed
+
+    """
+    if seed is None:
+        return _sg_random.randint(0, 2 ** 32 - 1)
+    else:
+        return seed
+
+
+def set_seed(seed, set_numpy=True, set_tensorflow=True, set_random=True):
```

Comment:
> I'm not sure what you meant here - wasn't your initial suggestion to not mutate the global seeds? I've switched to mutating a global random state that keeps generating seeds on the fly as @timpitman 's suggested - hopefully this resolves the weirdnesses you've mentioned?

Sorry, I mean consider the suggestion as not a good one anymore, not consider reapplying. English is ambiguous... 😄 

In particular, the value of a user to passing `set_numpy=False` here now seems low to me, compared to the risk of making a mistake. I think both of the following have to be true of the use-case for `set_numpy=False` (or one of the others) to be correct:

- explicitly don't want to set one of the seeds (for some reason?!)
- either don't need full reproducibility, or are sure the particular algorithms they're using don't use that library's randomness

As such, maybe we could just switch to always setting the seeds for all of them, and then even simplify slightly by removing the `_sg_random` value in favour of `random.randint(0, ...)` (so there's less risk of any threadsafety problems, although I suspect there are not any such problems at the moment).

In summary, maybe we could do:

```suggestion
def set_seed(seed):
```

What do you think? (Sorry for leading you up a garden path...)

Reply:
I was thinking of another use-case:

* the user wants to use some other utility of their own for managing seeds for numpy/tensorflow etc

I'm not sure how likely that is, but it felt like an easy option to include just in case. 

> simplify slightly by removing the _sg_random value in favour of random.randint(0, ...)

What do you mean by this? Don't we still need `_sg_random` to be a seedable `Random` object that acts as a seed generator?

Reply:
> the user wants to use some other utility of their own for managing seeds for numpy/tensorflow etc

Fair enough. Do utilities like these exist, or is it all adhoc code around `set_seed` similar to what we're doing here?

> Don't we still need _sg_random to be a seedable Random object that acts as a seed generator?
 
I don't think so: if we're seeding the global Python one, then `random.randint(0, 2**32 - 1)` will also give a reproducible sequence of seeds even when interleaved with other operations. I could well be misunderstanding what you're getting at?

Reply:
> Do utilities like these exist, or is it all adhoc code around set_seed similar to what we're doing here?

Not that I know of, I was thinking of just something adhoc like here

> if we're seeding the global Python one, then random.randint(0, 2**32 - 1) will also give a reproducible sequence of seeds even when interleaved with other operations

Ah I see what you mean I forgot we were setting the global python one already 😅 okay that does simplify things, will update 👍